man/lp.1
.\"
.\" lp man page for CUPS.
.\"
.\" Copyright © 2020-2024 by OpenPrinting.
.\" Copyright © 2007-2019 by Apple Inc.
.\" Copyright © 1997-2006 by Easy Software Products.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH lp 1 "CUPS" "2024-04-02" "OpenPrinting"
.SH NAME
lp \- print files
.SH SYNOPSIS
.B lp
[
\fB\-h \fIhostname\fR[\fB:\fIport\fR]
] [
.B \-E
] [
.B \-U
.I username
] [
.B \-c
] [
\fB\-d \fIdestination\fR[\fB/\fIinstance\fR]
] [
.B \-m
] [
.B \-n
.I num-copies
] [
\fB\-o \fIoption\fR[\fB=\fIvalue\fR]
] [
.B \-q
.I priority
] [
.B \-s
] [
.B \-t
.I title
] [
.B \-H
.I handling
] [
.B \-P
.I page-list
] [
.B \-\-
] [
.I file(s)
]
.br
.B lp
[
\fB\-h \fIhostname\fR[\fB:\fIport\fR]
] [
.B \-E
] [
.B \-U
.I username
] [
.B \-c
] [
.B \-i
.I job-id
] [
.B \-n
.I num-copies
] [
\fB\-o \fIoption\fR[\fB=\fIvalue\fR]
] [
.B \-q
.I priority
] [
.B \-t
.I title
] [
.B \-H
.I handling
] [
.B \-P
.I page-list
]
.SH DESCRIPTION
\fBlp\fR submits files for printing or alters a pending job.
Use a filename of "-" to force printing from the standard input.
.SS THE DEFAULT DESTINATION
CUPS provides many ways to set the default destination. The \fBLPDEST\fR and \fBPRINTER\fR environment variables are consulted first.
If neither are set, the current default set using the
.BR lpoptions (1)
command is used, followed by the default set using the
.BR lpadmin (8)
command.
.SH OPTIONS
The following options are recognized by \fIlp\fR:
.TP 5
.B \-\-
Marks the end of options; use this to print a file whose name begins with a dash (\-).
.TP 5
.B \-E
Forces encryption when connecting to the server.
.TP 5
\fB\-U \fIusername\fR
Specifies the username to use when connecting to the server.
.TP 5
.B \-c
This option is provided for backwards-compatibility only. On systems that support it, this option forces the print file to be copied to the spool directory before printing.
In CUPS, print files are always sent to the scheduler via IPP which has the same effect.
.TP 5
\fB\-d \fIdestination\fR
Prints files to the named printer.
.TP 5
\fB\-h \fIhostname\fR[\fB:\fIport\fR]
Chooses an alternate server.
Note: This option must occur before all others.
.TP 5
\fB\-i \fIjob-id\fR
Specifies an existing job to modify.
.TP 5
.B \-m
Sends an email when the job is completed.
.TP 5
\fB\-n \fIcopies\fR
Sets the number of copies to print.
.TP 5
\fB\-o "\fIname\fB=\fIvalue \fR[ ... \fIname\fB=\fIvalue \fR]\fB"\fR
Sets one or more job options.
See "COMMON JOB OPTIONS" below.
.TP 5
\fB\-q \fIpriority\fR
Sets the job priority from 1 (lowest) to 100 (highest).
The default priority is 50.
.TP 5
.B \-s
Do not report the resulting job IDs (silent mode.)
.TP 5
\fB\-t "\fIname\fB"\fR
Sets the job name.
.TP 5
\fB\-H \fIhh:mm\fR
.TP 5
\fB\-H hold\fR
.TP 5
\fB-H immediate\fR
.TP 5
\fB-H restart\fR
.TP 5
\fB-H resume\fR
Specifies when the job should be printed.
A value of \fIimmediate\fR will print the file immediately, a value of \fIhold\fR will hold the job indefinitely, and a UTC time value (HH:MM) will hold the job until the specified UTC (not local) time.
Use a value of \fIresume\fR with the \fI-i\fR option to resume a held job.
Use a value of \fIrestart\fR with the \fI-i\fR option to restart a completed job.
.TP 5
\fB\-P \fIpage-list\fR
Specifies which pages to print in the document.
The list can contain a list of numbers and ranges (#-#) separated by commas, e.g., "1,3-5,16".
The page numbers refer to the output pages and not the document's original pages - options like "number-up" can affect the numbering of the pages.
.SS COMMON JOB OPTIONS
Aside from the printer-specific options reported by the
.BR lpoptions (1)
command, the following generic options are available:
.TP 5
\fB\-o job-sheets=\fIname\fR\fR
Prints a cover page (banner) with the document.
The "name" can be "classified", "confidential", "secret", "standard", "topsecret", or "unclassified".
.TP 5
\fB\-o media=\fIsize\fR
Sets the page size to \fIsize\fR. Most printers support at least the size names "a4", "letter", and "legal".
.TP 5
\fB\-o number\-up=\fR{\fI2|4|6|8|9|16\fR}
Prints 2, 4, 6, 8, 9, or 16 document (input) pages on each output page.
.TP 5
\fB\-o orientation\-requested=4\fR
Prints the job in landscape (rotated 90 degrees counter-clockwise).
.TP 5
\fB\-o orientation\-requested=5\fR
Prints the job in landscape (rotated 90 degrees clockwise).
.TP 5
\fB\-o orientation\-requested=6\fR
Prints the job in reverse portrait (rotated 180 degrees).
.TP 5
\fB\-o print-as-raster\fR
Prints the job as raster.
.TP 5
\fB\-o print\-quality=3\fR
.TP 5
\fB\-o print\-quality=4\fR
.TP 5
\fB\-o print\-quality=5\fR
Specifies the output quality - draft (3), normal (4), or best (5).
.TP 5
\fB\-o sides=one\-sided\fR
Prints on one side of the paper.
.TP 5
\fB\-o sides=two\-sided\-long\-edge\fR
Prints on both sides of the paper for portrait output.
.TP 5
\fB\-o sides=two\-sided\-short\-edge\fR
Prints on both sides of the paper for landscape output.
.SH ENVIRONMENT
The following environment variables override the default destination:
.TP 5
.B LPDEST
Specifies the default print queue (System V standard).
.TP 5
.B PRINTER
Specifies the default print queue (Berkeley standard).
.SH CONFORMING TO
Unlike the System V printing system, CUPS allows printer names to contain any printable character except SPACE, TAB, "/", or "#".
Also, printer and class names are \fInot\fR case-sensitive.
.LP
The \fI-q\fR option accepts a different range of values than the Solaris lp command, matching the IPP job priority values (1-100, 100 is highest priority) instead of the Solaris values (0-39, 0 is highest priority).
.SH EXAMPLES
Print two copies of a document to the default printer:
.nf

    lp -n 2 filename

.fi
Print a double-sided legal document to a printer called "foo":
.nf

    lp -d foo -o media=legal -o sides=two-sided-long-edge filename

.fi
Print a presentation document 2-up to a printer called "bar":
.nf

    lp -d bar -o number-up=2 filename
.fi
.SH SEE ALSO
.BR cancel (1),
.BR lpadmin (8),
.BR lpoptions (1),
.BR lpq (1),
.BR lpr (1),
.BR lprm (1),
.BR lpstat (1),
CUPS Online Help (http://localhost:631/help)
.SH COPYRIGHT
Copyright \[co] 2020-2024 by OpenPrinting.


CHANGES.md
CHANGES - OpenPrinting CUPS
===========================


v2.4.20 - YYYY-MM-DD
--------------------

- CVE-2026-55453: The backend did not sanitize IPP attribute strings.
- CVE-2026-55467: The scheduler did not sanitize fax numbers.
- CVE-2026-55480: The scheduler did not open temporary PPD files in exclusive
  mode.
- CVE-2026-61702: The scheduler did not validate the request language value.
- CVE-2026-NNNNN: The `cupsUTF32toUTF8` function incorrectly treated UTF-32
  values as 64-bit.
- Increased the size of the SNMP supply name buffer used by the network backends
  (Issue #1604)
- The USB backend now clears a halt on USB errors (Issue #1606)
- Updated a few character tests for signed char platforms (Issue #1623)
- Added validation of IPP "printer-state-reasons" and
  "printer-mandatory-job-attributes" attributes (Issue #1632)
- Added missing Set-Printer-Attributes policy to cupsd.conf.
- Fixed mapping of standard PPD/PWG/legacy media size names to the local PPD
  size name (Issue #1375)
- Fixed handling of multiple PPD: keywords from filters (Issue #1562,
  related to CVE-2026-34980)
- Fixed handling of Kerberos user@REALM identities for user validation and
  quotas when StripUserDomain is not enabled (Issue #1584)
- Fixed raster error reporting overflow (Issue #1607)
- Fixed `cupsRasterInterpretPPD` handling of bad numbers (Issue #1608)
- Fixed SNMP hex string debug output (Issue #1610)
- Fixed some web interface bugs (Issue #1611)
- Fixed some compression issues in the rastertoepson and rastertohp drivers
  (Issue #1613)
- Fixed MIME `char` rule handling (Issue #1614)
- Fixed duplicate local printers (Issue #1531, Issue #1586, Issue #1593,
  Issue #1620)
- Fixed PPD cache memory leak (Issue #1629, Issue #1344)
- Fixed escaping of spaces in option values (Issue #1630)
- Fixed potential buffer overflow in `cupsCopyDestConflicts` (Issue #1631)
- Fixed backchannel parsing bug in `commandtops` filter (Issue #1637)
- Fixed section parsing in the web help indexing code (Issue #1641)
- Fixed potential buffer overrun in rastertolabel filter (Issue #1644)
- Fixed potential buffer overrun in rastertohp filter (Issue #1650)
- Fixed media selection with a mix of PPD and IPP options (Issue #1651)
- Fixed potential access of deleted IPP Everywhere printer (Issue #1655)
- Fixed limiting of PPD custom number output for large numbers (Issue #1656)
- Fixed a potential output length bug in the rastertohp driver (Issue #1658)
- Fixed a potential buffer underflow buf in the `ippAdd/SetStringf(v)` functions
  (Issue #1664)
- Fixed several issues reported by Coverity
- Fixed case-sensitive PPD keyword comparisons when filtering keyword updates
  from filters.
- Fixed potential buffer overrun in `cupsDoAuthentication`.


v2.4.19 - 2026-04-27
--------------------

- Fixed a regression in shared printing from non-local accounts (Issue #1557,
  related to CVE-2026-27447)


v2.4.18 - 2026-04-22
--------------------

- Fixed cupsd crash if user does not exist (Issue #1555, related to CVE-2026-27447)


v2.4.17 - 2026-04-17
--------------------

- CVE-2026-27447: The scheduler treated local user and group names as case-
  insensitive.
- CVE-2026-34978: The RSS notifier could write outside the scheduler's RSS
  directory.
- CVE-2026-34980: The scheduler did not filter control characters from option
  values.
- CVE-2026-34979: The scheduler did not always allocate enough memory for a
  job's options string.
- CVE-2026-34990: The scheduler incorrectly allowed local certificates over the
  loopback interface.
- CVE-2026-39314: Fixed the range check for job password strings.
- CVE-2026-39316: Fixed a printer subscription bug in the scheduler.
- CVE-2026-41079: Fixed a SNMP string conversion bug in the backends.
- The scheduler followed symbolic links when cleaning out its temporary
  directory (Issue #1448)
- Updated `cupsFileGetConf` and `cupsFilePutConf` to escape more characters.
- Updated man page `cancel` (Issue #984)
- Updated `cupsRasterReadHeader` to validate more of the page header values
  (Issue #1501)
- Fixed an issue with the class/printer CGI name checking.
- Fixed infinite loop in `http_write()` on busy print servers (Issue #827)
- Fixed potential TLS blocking issues (Issue #1128)
- Fixed a job history bug in the scheduler (Issue #1440)
- Fixed notifier logging bug that would result in nul bytes getting into the
  log (Issue #1450)
- Fixed possible use-after-free in `cupsdReadClient()` (Issue #1454)
- Fixed a document format bug in the IPP backend (Issue #1457)
- Fixed DRAIN_OUTPUT race condition (Issue #1461)
- Fixed a bug when then `ippFindXxx` and `ippSetXxx` functions were mixed.
- Fixed the mapping of supply type keywords to SNMP names.
- Fixed a bug in the IPP backend when SNMP was disabled.
- Fixed a crash bug in the rastertoepson filter.
- Fixed a bug in cgiCheckVariables.
- Fixed handling read/write errors with OpenSSL (Issue #1506)
- Fixed handling rehandshake error in `_httpTLSRead` (Issue #1508)
- Fixed a debug printf bug on Windows (Issue #1529)

systemv/lp.c
/*
 * "lp" command for CUPS.
 *
 * Copyright © 2020-2024 by OpenPrinting.
 * Copyright © 2007-2021 by Apple Inc.
 * Copyright © 1997-2007 by Easy Software Products.
 *
 * Licensed under Apache License v2.0.  See the file "LICENSE" for more
 * information.
 */

/*
 * Include necessary headers...
 */

#include <cups/cups-private.h>


/*
 * Local functions.
 */

static int	restart_job(const char *command, int job_id, const char *job_hold_until);
static int	set_job_attrs(const char *command, int job_id, int num_options, cups_option_t *options);
static void	usage(void) _CUPS_NORETURN;


/*
 * 'main()' - Parse options and send files for printing.
 */

int
main(int  argc,				/* I - Number of command-line arguments */
     char *argv[])			/* I - Command-line arguments */
{
  int		i, j;			/* Looping vars */
  int		job_id;			/* Job ID */
  char		*printer,		/* Printer name */
		*instance,		/* Instance name */
		*opt,			/* Option pointer */
		*val,			/* Option value */
		*title;			/* Job title */
  int		priority;		/* Job priority (1-100) */
  int		num_copies;		/* Number of copies per file */
  int		num_files;		/* Number of files to print */
  const char	*files[1000];		/* Files to print */
  cups_dest_t	*dest;			/* Selected destination */
  int		num_options;		/* Number of options */
  cups_option_t	*options;		/* Options */
  int		end_options;		/* No more options? */
  int		silent;			/* Silent or verbose output? */
  char		buffer[8192];		/* Copy buffer */


#ifdef __sun
 /*
  * Solaris does some rather strange things to re-queue remote print
  * jobs.  On bootup, the "lp" command is run as "printd" to re-spool
  * any remote jobs in /var/spool/print.  Since CUPS doesn't need this
  * nonsense, we just need to add the necessary check here to prevent
  * lp from causing boot problems...
  */

  if ((val = strrchr(argv[0], '/')) != NULL)
    val ++;
  else
    val = argv[0];

  if (!strcmp(val, "printd"))
    return (0);
#endif /* __sun */

  _cupsSetLocale(argv);

  silent      = 0;
  printer     = NULL;
  dest        = NULL;
  num_options = 0;
  options     = NULL;
  num_files   = 0;
  title       = NULL;
  job_id      = 0;
  end_options = 0;

  for (i = 1; i < argc; i ++)
  {
    if (!strcmp(argv[i], "--help"))
      usage();
    else if (argv[i][0] == '-' && argv[i][1] && !end_options)
    {
      for (opt = argv[i] + 1; *opt; opt ++)
      {
        switch (*opt)
	{
	  case 'E' : /* Encrypt */
	      cupsSetEncryption(HTTP_ENCRYPTION_REQUIRED);
	      break;

	  case 'U' : /* Username */
	      if (opt[1] != '\0')
	      {
		cupsSetUser(opt + 1);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;
		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected username after \"-U\" option."), argv[0]);
		  usage();
		}

		cupsSetUser(argv[i]);
	      }
	      break;

	  case 'c' : /* Copy to spool dir (always enabled) */
	      break;

	  case 'd' : /* Destination printer or class */
	      if (opt[1] != '\0')
	      {
		printer = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected destination after \"-d\" option."), argv[0]);
		  usage();
		}

		printer = argv[i];
	      }

	      if ((instance = strrchr(printer, '/')) != NULL)
		*instance++ = '\0';

	      if ((dest = cupsGetNamedDest(CUPS_HTTP_DEFAULT, printer,
					   instance)) != NULL)
	      {
		for (j = 0; j < dest->num_options; j ++)
		  if (cupsGetOption(dest->options[j].name, num_options,
				    options) == NULL)
		    num_options = cupsAddOption(dest->options[j].name,
						dest->options[j].value,
						num_options, &options);
	      }
	      else if (cupsGetError() == IPP_STATUS_ERROR_BAD_REQUEST ||
		       cupsGetError() == IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED)
	      {
		_cupsLangPrintf(stderr,
				_("%s: Error - add '/version=1.1' to server "
				  "name."), argv[0]);
		return (1);
	      }
	      else if (cupsGetError() == IPP_STATUS_ERROR_NOT_FOUND)
	      {
		_cupsLangPrintf(stderr,
				_("%s: Error - The printer or class does not exist."), argv[0]);
		return (1);
	      }
	      break;

	  case 'f' : /* Form */
	      if (opt[1] != '\0')
	      {
	        opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected form after \"-f\" option."), argv[0]);
		  usage();
		}
	      }

	      _cupsLangPrintf(stderr, _("%s: Warning - form option ignored."), argv[0]);
	      break;

	  case 'h' : /* Destination host */
	      if (opt[1] != '\0')
	      {
		cupsSetServer(opt + 1);
	        opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected hostname after \"-h\" option."), argv[0]);
		  usage();
		}

		cupsSetServer(argv[i]);
	      }
	      break;

	  case 'i' : /* Change job */
	      if (opt[1] != '\0')
	      {
		val = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Expected job ID after \"-i\" option."), argv[0]);
		  usage();
		}

		val = argv[i];
	      }

	      if (num_files > 0)
	      {
		_cupsLangPrintf(stderr, _("%s: Error - cannot print files and alter jobs simultaneously."), argv[0]);
		return (1);
	      }

	      if (strrchr(val, '-') != NULL)
		job_id = atoi(strrchr(val, '-') + 1);
	      else
		job_id = atoi(val);

	      if (job_id < 0)
	      {
		_cupsLangPrintf(stderr, _("%s: Error - bad job ID."), argv[0]);
		break;
	      }
	      break;

	  case 'm' : /* Send email when job is done */
#ifdef __sun
	  case 'p' : /* Notify on completion */
#endif /* __sun */
	  case 'w' : /* Write to console or email */
	      {
		char	email[1024];	/* EMail address */


		snprintf(email, sizeof(email), "mailto:%s@%s", cupsGetUser(), httpGetHostname(NULL, buffer, sizeof(buffer)));
		num_options = cupsAddOption("notify-recipient-uri", email, num_options, &options);
	      }

	      silent = 1;
	      break;

	  case 'n' : /* Number of copies */
	      if (opt[1] != '\0')
	      {
		num_copies = atoi(opt + 1);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected copies after \"-n\" option."), argv[0]);
		  usage();
		}

		num_copies = atoi(argv[i]);
	      }

	      if (num_copies < 1)
	      {
		_cupsLangPrintf(stderr, _("%s: Error - copies must be 1 or more."), argv[0]);
		return (1);
	      }

	      num_options = cupsAddIntegerOption("copies", num_copies, num_options, &options);
	      break;

	  case 'o' : /* Option */
	      if (opt[1] != '\0')
	      {
		num_options = cupsParseOptions(opt + 1, num_options, &options);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected option=value after \"-o\" option."), argv[0]);
		  usage();
		}

		num_options = cupsParseOptions(argv[i], num_options, &options);
	      }
	      break;

#ifndef __sun
	  case 'p' : /* Queue priority */
#endif /* !__sun */
	  case 'q' : /* Queue priority */
	      if (opt[1] != '\0')
	      {
		priority = atoi(opt + 1);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		if ((i + 1) >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected priority after \"-%c\" option."), argv[0], *opt);
		  usage();
		}

		i ++;

		priority = atoi(argv[i]);
	      }

	     /*
	      * For 100% Solaris compatibility, need to add:
	      *
	      *   priority = 99 * (39 - priority) / 39 + 1;
	      *
	      * However, to keep CUPS lp the same across all platforms
	      * we will break compatibility this far...
	      */

	      if (priority < 1 || priority > 100)
	      {
		_cupsLangPrintf(stderr, _("%s: Error - priority must be between 1 and 100."), argv[0]);
		return (1);
	      }

	      num_options = cupsAddIntegerOption("job-priority", priority, num_options, &options);
	      break;

	  case 's' : /* Silent */
	      silent = 1;
	      break;

	  case 't' : /* Title */
	      if (opt[1] != '\0')
	      {
		title = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected title after \"-t\" option."), argv[0]);
		  usage();
		}

		title = argv[i];
	      }
	      break;

	  case 'y' : /* mode-list */
	      if (opt[1] != '\0')
	      {
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected mode list after \"-y\" option."), argv[0]);
		  usage();
		}
	      }

	      _cupsLangPrintf(stderr, _("%s: Warning - mode option ignored."), argv[0]);
	      break;

	  case 'H' : /* Hold job */
	      if (opt[1] != '\0')
	      {
		val = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected hold name after \"-H\" option."), argv[0]);
		  usage();
		}

		val = argv[i];
	      }

	      if (!strcmp(val, "hold"))
		num_options = cupsAddOption("job-hold-until", "indefinite", num_options, &options);
	      else if (!strcmp(val, "resume") || !strcmp(val, "release"))
		num_options = cupsAddOption("job-hold-until", "no-hold", num_options, &options);
	      else if (!strcmp(val, "immediate"))
	      {
		num_options = cupsAddOption("job-hold-until", "no-hold", num_options, &options);
		num_options = cupsAddOption("job-priority", "100", num_options, &options);
	      }
	      else if (!strcmp(val, "restart"))
	      {
		if (job_id < 1)
		{
		  _cupsLangPrintf(stderr, _("%s: Need job ID (\"-i jobid\") before \"-H restart\"."), argv[0]);
		  return (1);
		}

		if (restart_job(argv[0], job_id, cupsGetOption("job-hold-until", num_options, options)))
		  return (1);
	      }
	      else
		num_options = cupsAddOption("job-hold-until", val, num_options, &options);
	      break;

	  case 'P' : /* Page list */
	      if (opt[1] != '\0')
	      {
		val = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected page list after \"-P\" option."), argv[0]);
		  usage();
		}

		val = argv[i];
	      }

	      num_options = cupsAddOption("page-ranges", val, num_options, &options);
	      break;

	  case 'S' : /* character set */
	      if (opt[1] != '\0')
	      {
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected character set after \"-S\" option."), argv[0]);
		  usage();
		}
	      }

	      _cupsLangPrintf(stderr, _("%s: Warning - character set option ignored."), argv[0]);
	      break;

	  case 'T' : /* Content-Type */
	      if (opt[1] != '\0')
	      {
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected content type after \"-T\" option."), argv[0]);
		  usage();
		}
	      }

	      _cupsLangPrintf(stderr, _("%s: Warning - content type option ignored."), argv[0]);
	      break;

	  case '-' : /* Stop processing options */
	      if (opt[1] != '\0')
	      {
		_cupsLangPrintf(stderr, _("%s: Error - unknown option \"%s\"."), argv[0], argv[i]);
		usage();
	      }

	      end_options = 1;
	      break;

	  default :
	      _cupsLangPrintf(stderr, _("%s: Error - unknown option \"%c\"."), argv[0], *opt);
	      usage();
	}
      }
    }
    else if (!strcmp(argv[i], "-"))
    {
      if (num_files || job_id)
      {
        _cupsLangPrintf(stderr,
			_("%s: Error - cannot print from stdin if files or a "
		          "job ID are provided."), argv[0]);
	return (1);
      }

      break;
    }
    else if (num_files < 1000 && job_id == 0)
    {
     /*
      * Print a file...
      */

      if (access(argv[i], R_OK) != 0)
      {
        _cupsLangPrintf(stderr, _("%s: Error - unable to access \"%s\" - %s"), argv[0], argv[i], strerror(errno));
        return (1);
      }

      files[num_files] = argv[i];
      num_files ++;

      if (title == NULL)
      {
        if ((title = strrchr(argv[i], '/')) != NULL)
	  title ++;
	else
          title = argv[i];
      }
    }
    else
    {
      _cupsLangPrintf(stderr, _("%s: Error - too many files - \"%s\"."), argv[0], argv[i]);
    }
  }

 /*
  * See if we are altering an existing job...
  */

  if (job_id)
    return (set_job_attrs(argv[0], job_id, num_options, options));

 /*
  * See if we have any files to print; if not, print from stdin...
  */

  if (printer == NULL)
  {
    if ((dest = cupsGetNamedDest(NULL, NULL, NULL)) != NULL)
    {
      printer = dest->name;

      for (j = 0; j < dest->num_options; j ++)
	if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
	  num_options = cupsAddOption(dest->options[j].name,
		                      dest->options[j].value,
				      num_options, &options);
    }
    else if (cupsGetError() == IPP_STATUS_ERROR_BAD_REQUEST ||
	     cupsGetError() == IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED)
    {
      _cupsLangPrintf(stderr,
		      _("%s: Error - add '/version=1.1' to server "
			"name."), argv[0]);
      return (1);
    }
  }

  if (printer == NULL)
  {
    if (!cupsGetNamedDest(NULL, NULL, NULL) && cupsGetError() == IPP_STATUS_ERROR_NOT_FOUND)
      _cupsLangPrintf(stderr, _("%s: Error - %s"), argv[0], cupsGetErrorString());
    else
      _cupsLangPrintf(stderr, _("%s: Error - scheduler not responding."), argv[0]);

    return (1);
  }

  if (num_files > 0)
    job_id = cupsPrintFiles(printer, num_files, files, title, num_options, options);
  else if ((job_id = cupsCreateJob(CUPS_HTTP_DEFAULT, printer,
                                   title ? title : "(stdin)",
                                   num_options, options)) > 0)
  {
    http_status_t	status;		/* Write status */
    const char		*format;	/* Document format */
    ssize_t		bytes;		/* Bytes read */

    if (cupsGetOption("raw", num_options, options))
      format = CUPS_FORMAT_RAW;
    else if ((format = cupsGetOption("document-format", num_options,
                                     options)) == NULL)
      format = CUPS_FORMAT_AUTO;

    status = cupsStartDocument(CUPS_HTTP_DEFAULT, printer, job_id, NULL,
                               format, 1);

    while (status == HTTP_STATUS_CONTINUE &&
           (bytes = read(0, buffer, sizeof(buffer))) > 0)
      status = cupsWriteRequestData(CUPS_HTTP_DEFAULT, buffer, (size_t)bytes);

    if (status != HTTP_STATUS_CONTINUE)
    {
      _cupsLangPrintf(stderr, _("%s: Error - unable to queue from stdin - %s."),
		      argv[0], httpStatus(status));
      cupsFinishDocument(CUPS_HTTP_DEFAULT, printer);
      cupsCancelJob2(CUPS_HTTP_DEFAULT, printer, job_id, 0);
      return (1);
    }

    if (cupsFinishDocument(CUPS_HTTP_DEFAULT, printer) != IPP_STATUS_OK)
    {
      _cupsLangPrintf(stderr, "%s: %s", argv[0], cupsGetErrorString());
      cupsCancelJob2(CUPS_HTTP_DEFAULT, printer, job_id, 0);
      return (1);
    }
  }

  if (job_id < 1)
  {
    _cupsLangPrintf(stderr, "%s: %s", argv[0], cupsGetErrorString());
    return (1);
  }
  else if (!silent)
    _cupsLangPrintf(stdout, _("request id is %s-%d (%d file(s))"),
		    printer, job_id, num_files);

  return (0);
}


/*
 * 'restart_job()' - Restart a job.
 */

static int				/* O - Exit status */
restart_job(const char *command,	/* I - Command name */
            int        job_id,		/* I - Job ID */
            const char *job_hold_until)	/* I - "job-hold-until" value, if any */
{
  ipp_t		*request;		/* IPP request */
  char		uri[HTTP_MAX_URI];	/* URI for job */


  request = ippNewRequest(IPP_OP_RESTART_JOB);

  snprintf(uri, sizeof(uri), "ipp://localhost/jobs/%d", job_id);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
               "job-uri", NULL, uri);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
               "requesting-user-name", NULL, cupsGetUser());

  if (job_hold_until)
    ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_KEYWORD, "job-hold-until", NULL, job_hold_until);

  ippDelete(cupsDoRequest(CUPS_HTTP_DEFAULT, request, "/jobs"));

  if (cupsGetError() == IPP_STATUS_ERROR_BAD_REQUEST ||
      cupsGetError() == IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED)
  {
    _cupsLangPrintf(stderr,
		    _("%s: Error - add '/version=1.1' to server "
		      "name."), command);
    return (1);
  }
  else if (cupsGetError() > IPP_STATUS_OK_CONFLICTING)
  {
    _cupsLangPrintf(stderr, "%s: %s", command, cupsGetErrorString());
    return (1);
  }

  return (0);
}


/*
 * 'set_job_attrs()' - Set job attributes.
 */

static int				/* O - Exit status */
set_job_attrs(
    const char    *command,		/* I - Command name */
    int           job_id,		/* I - Job ID */
    int           num_options,		/* I - Number of options */
    cups_option_t *options)		/* I - Options */
{
  ipp_t		*request;		/* IPP request */
  char		uri[HTTP_MAX_URI];	/* URI for job */


  if (num_options == 0)
    return (0);

  request = ippNewRequest(IPP_OP_SET_JOB_ATTRIBUTES);

  snprintf(uri, sizeof(uri), "ipp://localhost/jobs/%d", job_id);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
               "job-uri", NULL, uri);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
               "requesting-user-name", NULL, cupsGetUser());

  cupsEncodeOptions(request, num_options, options);

  ippDelete(cupsDoRequest(CUPS_HTTP_DEFAULT, request, "/jobs"));

  if (cupsGetError() == IPP_STATUS_ERROR_BAD_REQUEST ||
      cupsGetError() == IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED)
  {
    _cupsLangPrintf(stderr,
		    _("%s: Error - add '/version=1.1' to server "
		      "name."), command);
    return (1);
  }
  else if (cupsGetError() > IPP_STATUS_OK_CONFLICTING)
  {
    _cupsLangPrintf(stderr, "%s: %s", command, cupsGetErrorString());
    return (1);
  }

  return (0);
}


/*
 * 'usage()' - Show program usage and exit.
 */

static void
usage(void)
{
  _cupsLangPuts(stdout, _("Usage: lp [options] [--] [file(s)]\n"
                          "       lp [options] -i id"));
  _cupsLangPuts(stdout, _("Options:"));
  _cupsLangPuts(stdout, _("-c                      Make a copy of the print file(s)"));
  _cupsLangPuts(stdout, _("-d destination          Specify the destination"));
  _cupsLangPuts(stdout, _("-E                      Encrypt the connection to the server"));
  _cupsLangPuts(stdout, _("-h server[:port]        Connect to the named server and port"));
  _cupsLangPuts(stdout, _("-H HH:MM                Hold the job until the specified UTC time"));
  _cupsLangPuts(stdout, _("-H hold                 Hold the job until released/resumed"));
  _cupsLangPuts(stdout, _("-H immediate            Print the job as soon as possible"));
  _cupsLangPuts(stdout, _("-H restart              Reprint the job"));
  _cupsLangPuts(stdout, _("-H resume               Resume a held job"));
  _cupsLangPuts(stdout, _("-i id                   Specify an existing job ID to modify"));
  _cupsLangPuts(stdout, _("-m                      Send an email notification when the job completes"));
  _cupsLangPuts(stdout, _("-n num-copies           Specify the number of copies to print"));
  _cupsLangPuts(stdout, _("-o option[=value]       Specify a printer-specific option"));
  _cupsLangPuts(stdout, _("-o job-sheets=standard  Print a banner page with the job"));
  _cupsLangPuts(stdout, _("-o media=size           Specify the media size to use"));
  _cupsLangPuts(stdout, _("-o number-up=N          Specify that input pages should be printed N-up (1, 2, 4, 6, 8, 9, and 16 are supported)"));
  _cupsLangPuts(stdout, _("-o orientation-requested=N\n"
                          "                        Specify portrait (3) or landscape (4) orientation"));
  _cupsLangPuts(stdout, _("-o print-quality=N      Specify the print quality - draft (3), normal (4), or best (5)"));
  _cupsLangPuts(stdout, _("-o sides=one-sided      Specify 1-sided printing"));
  _cupsLangPuts(stdout, _("-o sides=two-sided-long-edge\n"
                          "                        Specify 2-sided portrait printing"));
  _cupsLangPuts(stdout, _("-o sides=two-sided-short-edge\n"
                          "                        Specify 2-sided landscape printing"));
  _cupsLangPuts(stdout, _("-P page-list            Specify a list of pages to print"));
  _cupsLangPuts(stdout, _("-q priority             Specify the priority from low (1) to high (100)"));
  _cupsLangPuts(stdout, _("-s                      Be silent"));
  _cupsLangPuts(stdout, _("-t title                Specify the job title"));
  _cupsLangPuts(stdout, _("-U username             Specify the username to use for authentication"));


  exit(1);
}
